modules/extensions-support/src/main/java/org/mule/module/extension/internal/runtime/metadata/MetadataMediator.java

```diff
@@ -0,0 +1,269 @@
+/*
+ * Copyright (c) MuleSoft, Inc.  All rights reserved.  http://www.mulesoft.com
+ * The software in this package is published under the terms of the CPAL v1.0
+ * license, a copy of which has been included with this distribution in the
+ * LICENSE.txt file.
+ */
+package org.mule.module.extension.internal.runtime.metadata;
+
+import static org.mule.api.metadata.FailureType.NO_DYNAMIC_TYPE_AVAILABLE;
+import static org.mule.util.metadata.ResultFactory.failure;
+import static org.mule.util.metadata.ResultFactory.mergeResults;
+import static org.mule.util.metadata.ResultFactory.success;
+import org.mule.api.metadata.MetadataKey;
+import org.mule.api.metadata.Result;
+import org.mule.api.metadata.descriptor.ImmutableOperationMetadataDescriptor;
+import org.mule.api.metadata.descriptor.ImmutableParameterMetadataDescriptor;
+import org.mule.api.metadata.descriptor.OperationMetadataDescriptor;
+import org.mule.api.metadata.descriptor.ParameterMetadataDescriptor;
+import org.mule.api.temporary.MuleMessage;
+import org.mule.extension.api.annotation.param.metadata.Content;
+import org.mule.extension.api.annotation.param.metadata.MetadataKeyParam;
+import org.mule.extension.api.introspection.OperationModel;
+import org.mule.extension.api.introspection.ParameterModel;
+import org.mule.extension.api.introspection.metadata.MetadataResolverFactory;
+import org.mule.extension.api.metadata.MetadataContext;
+import org.mule.extension.api.metadata.MetadataResolver;
+import org.mule.extension.api.metadata.MetadataResolvingException;
+import org.mule.extension.api.metadata.NullMetadataKey;
+import org.mule.metadata.api.model.MetadataType;
+import org.mule.metadata.api.model.NullType;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public class MetadataMediator
```

Comment:
I would expect this class to be final and with a reduced visibility (might not be possible).

I will skip review of this class for now since I know that prior comments will lead to a major rewrite of it